modules/diskq/dqtool.c

```diff
@@ -241,12 +251,292 @@ _relocate_validate_options(void)
   return _validate_diskq_path(new_diskq_path) && _validate_persist_file_path(persist_file_path);
 }
 
+static gint
+_open_fd_for_reading(const gchar *fname)
+{
+  gint fd = open(fname, O_RDONLY);
+
+  if (fd == -1)
+    fprintf(stderr, "Failed to open %s, reason: %s\n", fname, strerror(errno));
+
+  return fd;
+}
+
+static gboolean
+_get_content_length_by_fd(gint fd, gsize *size)
+{
+  struct stat st;
+
+  if (fstat(fd, &st) == -1)
+    {
+      fprintf(stderr, "Error: %s\n", strerror(errno));
+      *size = 0;
+      return FALSE;
+    }
+
+  *size = st.st_size;
+  return TRUE;
+}
+
+static gboolean
+_mmap_content_from_fd(gint fd, gchar **content, gsize *content_length)
+{
+  if (!_get_content_length_by_fd(fd, content_length))
+    return FALSE;
+
+  *content = mmap(NULL, *content_length, PROT_READ, MAP_PRIVATE, fd, 0);
```

Comment:
I feel some risk here. Disk queue files might be large. Can we run into trouble mmaping gigabytes of ram? (I am thinking about ulimit).

Reply:
`mmap` is an easy&good alternative for fread/fwrite, but well.. if someone limiting the address space (with `ulimit -v`) then this won't work (@Kokan : but it works even when you have a small amount of memory... so if you have 3GB of RAM, you can easily load a 6GB diskq file - and use a 12GB virt. address space).

`g_file_copy`: that uses gio (and currently we don't link to gio lib)
(or, `g_file_move`)

So if you don't like the `mmap` based copy, I can re-implement it to fread/fwrite, but won't/cannot use `g_file_copy`.